python/ray/rllib/utils/torch_policy_graph.py

```diff
@@ -0,0 +1,106 @@
+            action_space (gym.Space): action space of the policy.
+            model (nn.Module): PyTorch policy module. Given observations as
+                input, this module must a list of outputs where the first item
+                are action logits, and the remainder can be any value.
+            loss (nn.Module): Loss defined as a PyTorch module. The inputs for
+                this module are defined by the `loss_inputs` param. This module
+                returns a single scalar loss.
+            loss_inputs (list): List of SampleBatch columns that will be
+                passed to the loss module's forward() function when computing
+                the loss. For example, ["obs", "action", "advantages"].
+
+        Attributes:
+            observation_space (gym.Space): observation space of the policy.
+            action_space (gym.Space): action space of the policy.
+        """
+        self.observation_space = observation_space
+        self.action_space = action_space
+        self.lock = Lock()
+        self._model = model
+        self._loss = loss
+        self._loss_inputs = loss_inputs
+        self._optimizer = self.optimizer()
+
+    def extra_action_out(self, model_out):
+        """Returns dict of extra info to include in experience batch.
+
+        Arguments:
+            model_out (list): Outputs of the policy model module."""
+        return {}
+
+    def optimizer(self):
+        """Custom PyTorch optimizer to use."""
+        return torch.optim.Adam(self._model.parameters())
+
+    def compute_actions(
+            self, obs_batch, state_batches=None, is_training=False):
+        if state_batches:
+            raise NotImplementedError("Torch RNN support")
+        with self.lock:
+            ob = torch.from_numpy(np.array(obs_batch)).float()
+            model_out = self._model(ob)
+            logits = model_out[0]  # assume the first output is the logits
+            actions = F.softmax(logits, dim=1).multinomial(1).squeeze(0)
+            return var_to_np(actions), [], self.extra_action_out(model_out)
+
+    def compute_gradients(self, postprocessed_batch):
+        with self.lock:
+            loss_in = []
+            for key in self._loss_inputs:
+                loss_in.append(torch.from_numpy(postprocessed_batch[key]))
+            loss_in = [
+                t.float() if isinstance(t, torch.DoubleTensor) else t
+                for t in loss_in]
+            loss_out = self._loss(*loss_in)
+            self._optimizer.zero_grad()
+            loss_out.backward()
+            # Note that return values are just references;
+            # calling zero_grad will modify the values
+            return [p.grad.data.numpy() for p in self._model.parameters()], {}
+
+    def apply_gradients(self, gradients):
+        with self.lock:
+            self._optimizer.zero_grad()
+            for g, p in zip(gradients, self._model.parameters()):
+                p.grad = torch.from_numpy(g)
```

Comment:
corresponding if not p.grad continue here. then you also wanto to copy_ to the data I think since you're feeding flat grads

Reply:
Do you mean on `g`?

Reply:
well, as mentioned above just assuming p.grad is not NOne is probably fine.

then just do p.grad.data.copy_(torch.from_numpy(g))

Reply:
Hm yeah maybe we can assume not None for now. Btw when I tried copy_() sometimes p.grad.data was None and that failed.